File: android_java_hummus/src/main/java/com/github/klee0kai/hummus/adapterdelegates/diffutil/SameDiffUtilHelper.java
package com.github.klee0kai.hummus.adapterdelegates.diffutil;

import androidx.recyclerview.widget.DiffUtil;

import com.github.klee0kai.hummus.model.CloneableHelper;
import com.github.klee0kai.hummus.model.ICloneable;

import java.util.ArrayList;
import java.util.List;

public class SameDiffUtilHelper<T extends ICloneable> {

    public static final int MAX_LIST_SIZE = 10_000;

    private List<T> oldList = null;
    private ListDiffResult<T> diffResult = null;

    public void saveOld(List<T> old, boolean deepCopy) {
        if (old == null) {
            this.oldList = null;
            return;
        }
        int len = old.size();
        if (len > MAX_LIST_SIZE) {
            //big list size for diff util
            return;
        }
        this.oldList = new ArrayList<>(len);
        if (deepCopy) {
            for (T item : old) {
                oldList.add(CloneableHelper.tryClone(item, item));
            }
        } else {
            this.oldList.addAll(old);
        }
    }

    public ListDiffResult<T> calculateWith(List<T> newList, boolean detectMoves) {
        if (oldList == null) {
            // nothing to compare
            return diffResult = new ListDiffResult<>(null, null, newList);
        }

        DiffUtil.DiffResult result = DiffUtil.calculateDiff(
                new SameDiffCallback(
                        new ArrayList<>(oldList),
                        newList != null ? new ArrayList<>(newList) : null,
                        false
                ),
                detectMoves);
        diffResult = new ListDiffResult<>(result, oldList, newList);
        oldList = null;
        return diffResult;
    }

    public ListDiffResult<T> calculateWith(List<T> newList) {
        return calculateWith(newList, false);
    }


    public ListDiffResult<T> popDiffResult(List<T> list) {
        ListDiffResult<T> changes = diffResult;
        diffResult = null;
        return changes != null ? changes : new ListDiffResult<>(null, null, list);
    }


}

File: android_kotlin_hummus/src/main/kotlin/com/github/klee0kai/hummus/adapterdelegates/diffutil/SameDiffUtilHelper.kt
package com.github.klee0kai.hummus.adapterdelegates.diffutil

import androidx.recyclerview.widget.DiffUtil

open class SameDiffUtilHelper<T : Any> {

    protected open var oldList: MutableList<T>? = null
    protected open var diffResult: ListDiffResult<T>? = null

    fun saveOld(old: List<T>?) {
        if (old == null) {
            oldList = null
            return
        }
        val len = old.size
        if (len > MAX_LIST_SIZE) {
            //big list size for diff util
            return
        }


        // we not prefer use deep copy for kotlin
        // use immutable data classes to protect data and for correct diff util working
        oldList = mutableListOf()
        oldList?.addAll(old)
    }

    fun calculateWith(newList: List<T>?, detectMoves: Boolean = false): ListDiffResult<T> {
        if (oldList == null) {
            // nothing to compare
            return ListDiffResult(null, null, newList).also { diffResult = it }
        }
        val result: DiffUtil.DiffResult = DiffUtil.calculateDiff(
            SameDiffCallback(
                oldList, newList, false
            ),
            detectMoves
        )
        diffResult = ListDiffResult(result, oldList, newList)
        oldList = null
        return diffResult!!
    }

    fun popDiffResult(list: List<T>?): ListDiffResult<T> {
        val changes = diffResult
        diffResult = null
        return changes ?: ListDiffResult(null, null, list)
    }

    companion object {
        private const val MAX_LIST_SIZE = 10000
    }
}